journalbeat/checkpoint/checkpoint.go

```diff
@@ -0,0 +1,290 @@
+	}
+
+	// Minimum flush interval.
+	if c.flushInterval < time.Second {
+		c.flushInterval = time.Second
+	}
+
+	// Read existing state information:
+	ps, err := c.read()
+	if err != nil {
+		return nil, err
+	}
+
+	if ps != nil {
+		for _, state := range ps.States {
+			c.states[state.Path] = state
+		}
+	}
+
+	// Write the state file to verify we have have permissions.
+	err = c.flush()
+	if err != nil {
+		return nil, err
+	}
+
+	c.wg.Add(1)
+	go c.run()
+	return c, nil
+}
+
+// run is worker loop that reads incoming state information from the save
+// channel and persists it when the number of changes reaches maxEvents or
+// the amount of time since the last disk write reaches flushInterval.
+func (c *Checkpoint) run() {
+	defer c.wg.Done()
+	defer c.persist()
+
+	flushTimer := time.NewTimer(c.flushInterval)
+	defer flushTimer.Stop()
+loop:
+	for {
+		select {
+		case <-c.done:
+			break loop
+		case s := <-c.save:
+			c.lock.Lock()
+			c.states[s.Path] = s
+			c.lock.Unlock()
+			c.numUpdates++
+			if c.numUpdates < c.maxUpdates {
+				continue
+			}
+		case <-flushTimer.C:
+		}
+
+		c.persist()
+		flushTimer.Reset(c.flushInterval)
+	}
+}
+
+// Shutdown stops the checkpoint worker (which persists any state to disk as
+// it stops). This method blocks until the checkpoint worker shutdowns. Calling
+// this method more once is safe and has no effect.
+func (c *Checkpoint) Shutdown() {
+	c.once.Do(func() {
+		close(c.done)
+		c.wg.Wait()
+	})
+}
+
+// States returns the current in-memory event log state. This state information
+// is bootstrapped with any data found on disk at creation time.
+func (c *Checkpoint) States() map[string]JournalState {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	copy := make(map[string]JournalState)
+	for k, v := range c.states {
+		copy[k] = v
+	}
+
+	return copy
+}
+
+// Persist queues the given event log state information to be written to disk.
+func (c *Checkpoint) Persist(path, cursor string, realTs, monotonicTs uint64) {
+	c.PersistState(JournalState{
+		Path:               path,
+		Cursor:             cursor,
+		RealtimeTimestamp:  realTs,
+		MonotonicTimestamp: monotonicTs,
+	})
+}
+
+// PersistState queues the given event log state to be written to disk.
+func (c *Checkpoint) PersistState(st JournalState) {
+	c.save <- st
+}
+
+// persist writes the current state to disk if the in-memory state is dirty.
+func (c *Checkpoint) persist() bool {
+	if c.numUpdates == 0 {
+		return false
+	}
+
+	err := c.flush()
+	if err != nil {
+		return false
+	}
+
+	logp.Debug("checkpoint", "Checkpoint saved to disk. numUpdates=%d",
+		c.numUpdates)
+	c.numUpdates = 0
+	return true
+}
+
+// flush writes the current state to disk.
+func (c *Checkpoint) flush() error {
+	c.fileLock.Lock()
+	defer c.fileLock.Unlock()
+
+	tempFile := c.file + ".new"
+	file, err := create(tempFile)
+	if os.IsNotExist(err) {
+		// Try to create directory if it does not exist.
+		if createDirErr := c.createDir(); createDirErr == nil {
```

Comment:
Is this something we should already try on startup? I "think" we do it in Filebeat to warn the user that he might miss permissions to update / create the registry file as in the past we had the issue that FB was running but states could not be written.